zio-http/src/main/scala/zio/http/ZClient.scala

```diff
@@ -633,25 +637,39 @@ object ZClient {
                 }.tapErrorCause { cause =>
                   channelScope.close(Exit.failCause(cause))
                 }
+
+                _ <- onResponse.await.onInterrupt {
+                  // propagate the interrupt to onComplete
+                  onComplete.interrupt
+                }.forkDaemon
+
                 // If request registration failed we release the channel immediately.
                 // Otherwise we wait for completion signal from netty in a background fiber:
-                _            <-
+                _ <-
                   onComplete.await.interruptible.exit.flatMap { exit =>
-                    resetChannel
-                      .zip(exit)
-                      .map { case (s1, s2) => s1 && s2 }
-                      .catchAll(_ =>
-                        ZIO.succeed(ChannelState.Invalid),
-                      ) // In case resetting the channel fails we cannot reuse it
-                      .flatMap { channelState =>
-                        connectionPool
-                          .invalidate(channel)
-                          .when(channelState == ChannelState.Invalid)
-                      }
-                      .flatMap { _ =>
-                        channelScope.close(exit)
-                      }
-                      .uninterruptible
+                    if (exit.isInterrupted) {
+                      // do not try to reset the channel because it is disconnected
```

Comment:
Should not we invalidate the channel in the connection pool in case of interrupt? 
If it is not reset it is not reusable for the next connection - this should work the same as the case when reset fails below, I think.

Reply:
I'm not sure. I was hoping there'd be some tests that validate the connection pool.  But maybe not. If the pipeline is disconnected, can the connection be reused? I don't fully understand how all this works.

Reply:
Unfortunately the existing connection pool tests are not covering interruption. It would be nice to have tests for it, for the full matrix of possibilities (streaming/non-streaming, http 1.1/1.0)

When the request finishes normally, there are some cleanup steps to be done on the netty pipeline - some handlers that are dynamically added during handling the request need to be removed, while the ones that are "static" for the connection has to remain there. I think when interrupting we cannot guarantee that this state is restored. Or maybe we can but then we need to explicitly do it.

Reply:
The reason I cannot give 100% straight answer is that even though I wrote that connection pool code I don't fully remember how it works anymore, need to refresh my understanding of Netty :)